[MYSQL] Caused by: java.lang.ArrayIndexOutOfBoundsException

**Describe the bug**
A clear and concise description of what the bug is.
java.lang.RuntimeException: One or more fetchers have encountered exception
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcherManager.checkErrors(SplitFetcherManager.java:199)
	at org.apache.flink.connector.base.source.reader.SourceReaderBase.getNextFetch(SourceReaderBase.java:154)
	at org.apache.flink.connector.base.source.reader.SourceReaderBase.pollNext(SourceReaderBase.java:116)
	at org.apache.flink.connector.base.source.reader.SourceReaderBase.pollNext(SourceReaderBase.java:141)
	at org.apache.flink.streaming.api.operators.SourceOperator.emitNext(SourceOperator.java:294)
	at org.apache.flink.streaming.runtime.io.StreamTaskSourceInput.emitNext(StreamTaskSourceInput.java:69)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:66)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:423)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:204)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:681)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.executeInvoke(StreamTask.java:636)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runWithCleanUpOnFail(StreamTask.java:647)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:620)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:779)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:566)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: SplitFetcher thread 0 received unexpected exception while polling the records
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:146)
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.run(SplitFetcher.java:101)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.errors.ConnectException: An exception occurred in the change event producer. This connector will be stopped.
	at io.debezium.pipeline.ErrorHandler.setProducerThrowable(ErrorHandler.java:42)
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$ReaderThreadLifecycleListener.onEventDeserializationFailure(MySqlStreamingChangeEventSource.java:1189)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:958)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:606)
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:850)
	... 1 more
Caused by: io.debezium.DebeziumException
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1142)
	... 5 more
Caused by: java.lang.ArrayIndexOutOfBoundsException
	at com.github.shyiko.mysql.binlog.io.BufferedSocketInputStream.read(BufferedSocketInputStream.java:65)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readWithinBlockBoundaries(ByteArrayInputStream.java:262)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:241)
	at java.io.InputStream.skip(InputStream.java:224)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.skipToTheEndOfTheBlock(ByteArrayInputStream.java:280)
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:305)
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:232)
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$1.nextEvent(MySqlStreamingChangeEventSource.java:233)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:945)
	... 3 more



**Environment :**
 - Flink version :  1.13.2
 - Flink CDC version: 2.0.1
 - Database and version: sql7

**To Reproduce**
Steps to reproduce the behavior:
1. Thes test data :
2. The test code :
3. The error :

**Additional Description**
If applicable, add screenshots to help explain your problem.

> could you try 2.0.2 version? @hwzp99

2.0.2，the same problem..  @leonardBang 
2021-09-24 13:39:45,947 ERROR io.debezium.pipeline.ErrorHandler                            [] - Producer failure
io.debezium.DebeziumException: null
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1142) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$ReaderThreadLifecycleListener.onEventDeserializationFailure(MySqlStreamingChangeEventSource.java:1189) [darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:958) [darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:606) [darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:850) [darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]
Caused by: java.lang.ArrayIndexOutOfBoundsException
	at com.github.shyiko.mysql.binlog.io.BufferedSocketInputStream.read(BufferedSocketInputStream.java:65) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readWithinBlockBoundaries(ByteArrayInputStream.java:262) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:241) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at java.io.InputStream.skip(InputStream.java:224) ~[?:1.8.0_202]
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.skipToTheEndOfTheBlock(ByteArrayInputStream.java:280) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:305) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:232) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$1.nextEvent(MySqlStreamingChangeEventSource.java:233) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:945) ~[darwinKafkaDataToHive-1.0-SNAPSHOT.jar:?]
	... 3 more

--- a/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
@@ -24,8 +24,10 @@ import java.io.InputStream;
 
 /**
  * Copied from https://github.com/osheroff/mysql-binlog-connector-java project to fix
- * https://github.com/ververica/flink-cdc-connectors/issues/460. We should remove this class once
- * https://github.com/osheroff/mysql-binlog-connector-java/issues/66 has been fixed
+ * https://github.com/ververica/flink-cdc-connectors/issues/460.
+ *
+ * <p>We should remove this class after we bumped a higher mysql-binlog-connector-java version where
+ * the https://github.com/osheroff/mysql-binlog-connector-java/issues/66 has been fixed.
  */
 public class BufferedSocketInputStream extends FilterInputStream {
 

--- a/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
@@ -26,8 +26,9 @@ import java.io.InputStream;
  * Copied from https://github.com/osheroff/mysql-binlog-connector-java project to fix
  * https://github.com/ververica/flink-cdc-connectors/issues/460.
  *
- * <p>We should remove this class after we bumped a higher mysql-binlog-connector-java version where
- * the https://github.com/osheroff/mysql-binlog-connector-java/issues/66 has been fixed.
+ * <p>Line 50 and Line 70 ~ 72: Returns -1 means reach the end of InputStream. We should remove this
+ * class after we bumped a higher mysql-binlog-connector-java version where the
+ * https://github.com/osheroff/mysql-binlog-connector-java/issues/66 has been fixed.
  */
 public class BufferedSocketInputStream extends FilterInputStream {
 
@@ -46,7 +47,7 @@ public class BufferedSocketInputStream extends FilterInputStream {
 
     @Override
     public int available() throws IOException {
-        return limit - offset + in.available();
+        return limit == -1 ? in.available() : limit - offset + in.available();
     }
 
     @Override

--- a/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
+++ b/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java
@@ -1,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.github.shyiko.mysql.binlog.io;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Copied from https://github.com/osheroff/mysql-binlog-connector-java project to fix
+ * https://github.com/ververica/flink-cdc-connectors/issues/460. We should remove this class once
+ * https://github.com/osheroff/mysql-binlog-connector-java/issues/66 has been fixed
+ */
+public class BufferedSocketInputStream extends FilterInputStream {
+
+    private byte[] buffer;
+    private int offset;
+    private int limit;
+
+    public BufferedSocketInputStream(InputStream in) {
+        this(in, 512 * 1024);
+    }
+
+    public BufferedSocketInputStream(InputStream in, int bufferSize) {
+        super(in);
+        this.buffer = new byte[bufferSize];
+    }
+
+    @Override
+    public int available() throws IOException {
+        return limit - offset + in.available();
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (offset < limit) {
+            return buffer[offset++] & 0xff;
+        }
+        offset = 0;
+        limit = in.read(buffer, 0, buffer.length);
+        return limit != -1 ? buffer[offset++] & 0xff : -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        if (offset >= limit) {
+            if (len >= buffer.length) {
+                return in.read(b, off, len);
+            }
+            offset = 0;
+            limit = in.read(buffer, 0, buffer.length);
+            if (limit == -1) {
+                return limit;
+            }
+        }
+        int bytesRemainingInBuffer = Math.min(len, limit - offset);
+        System.arraycopy(buffer, offset, b, off, bytesRemainingInBuffer);
+        offset += bytesRemainingInBuffer;
+        return bytesRemainingInBuffer;
+    }
+}

--- a/flink-connector-mysql-cdc/src/test/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStreamTest.java
+++ b/flink-connector-mysql-cdc/src/test/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStreamTest.java
@@ -1,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.github.shyiko.mysql.binlog.io;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/** Unit test for {@link BufferedSocketInputStream}. */
+public class BufferedSocketInputStreamTest {
+
+    @Test
+    public void testReadFromBufferedSocketInputStream() throws Exception {
+        BufferedSocketInputStream in =
+                new BufferedSocketInputStream(
+                        new ByteArrayInputStream(
+                                new byte[] {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'}));
+        byte[] buf = new byte[3];
+        assertEquals(3, in.read(buf, 0, buf.length));
+        assertCharBytesEquals(new byte[] {'A', 'B', 'C'}, buf);
+
+        assertEquals(3, in.read(buf, 0, buf.length));
+        assertCharBytesEquals(new byte[] {'D', 'E', 'F'}, buf);
+
+        assertEquals(2, in.read(buf, 0, buf.length));
+        assertCharBytesEquals(new byte[] {'G', 'H'}, buf);
+
+        // reach the end of stream normally
+        assertEquals(-1, in.read(buf, 0, buf.length));
+    }
+
+    private static void assertCharBytesEquals(byte[] expected, byte[] actual) {
+        for (int i = 0; i < expected.length; i++) {
+            assertEquals(expected[i], actual[i]);
+        }
+    }
+}

--- a/flink-connector-mysql-cdc/src/test/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStreamTest.java
+++ b/flink-connector-mysql-cdc/src/test/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStreamTest.java
@@ -20,6 +20,8 @@ package com.github.shyiko.mysql.binlog.io;
 
 import org.junit.Test;
 
+import java.util.Arrays;
+
 import static org.junit.Assert.assertEquals;
 
 /** Unit test for {@link BufferedSocketInputStream}. */
@@ -33,21 +35,19 @@ public class BufferedSocketInputStreamTest {
                                 new byte[] {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'}));
         byte[] buf = new byte[3];
         assertEquals(3, in.read(buf, 0, buf.length));
-        assertCharBytesEquals(new byte[] {'A', 'B', 'C'}, buf);
+        Arrays.equals(new byte[] {'A', 'B', 'C'}, buf);
+        assertEquals(5, in.available());
 
         assertEquals(3, in.read(buf, 0, buf.length));
-        assertCharBytesEquals(new byte[] {'D', 'E', 'F'}, buf);
+        Arrays.equals(new byte[] {'D', 'E', 'F'}, buf);
+        assertEquals(2, in.available());
 
         assertEquals(2, in.read(buf, 0, buf.length));
-        assertCharBytesEquals(new byte[] {'G', 'H'}, buf);
+        Arrays.equals(new byte[] {'G', 'H'}, buf);
+        assertEquals(0, in.available());
 
         // reach the end of stream normally
         assertEquals(-1, in.read(buf, 0, buf.length));
-    }
-
-    private static void assertCharBytesEquals(byte[] expected, byte[] actual) {
-        for (int i = 0; i < expected.length; i++) {
-            assertEquals(expected[i], actual[i]);
-        }
+        assertEquals(0, in.available());
     }
 }

